### ISSUE TITLE
When first adding a bot to Microsoft Team, the bot's TeamsInfo is incomplete

### ISSUE THREAD

**Issue by apeckstl** [microsoft/botbuilder-dotnet#3276]
## Version
Version 4.6

## Describe the bug
When first adding a bot to Microsoft Team, the bot's TeamsInfo only includes my account, even though there are several members in the team. I am using the TeamsInfo.GetMembersAsync method to get all of the members. I am calling the method in my overridden OnTeamsMembersAddedAsync, which I know gets called when the bot first gets added. Later I will add code to only call this when the bot is added. However, logging that list of members only returns my name. 

## To Reproduce
1. Using the Bot Builder sample 57, at https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/57.teams-conversation-bot
2. Modify the OnTeamsMembersAddedAsync():
```
protected override async Task OnTeamsMembersAddedAsync(IList<TeamsChannelAccount> membersAdded, TeamInfo teamInfo, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
{
    IEnumerable<TeamsChannelAccount> members = await TeamsInfo.GetMembersAsync(turnContext, cancellationToken);
    foreach (var member in members)
    {
        Logger.LogInformation(member.Name);
    }
}
```
3. Add the bot to a Team (I added it to the general channel)
4. In logging output, only my name is logged. 

## Expected behavior
I expect a full list of the members in the team.

## Additional context
I have also tried sending a message to the turnContext passed into the OnTeamsMembersAddedAsync method, but it just sends the message directly to me through chat, even though the bot was added to a team. This makes me think the bot is acting as though I just added it to a personal context, but every subsequent activity the bot receives is treated as a teams context.

[bug]

**Comment by Virtual-Josh** [comment]
Hi @apeckstl,

A couple of things. 

First, you will only ever get the bot in the members added event the very first time that you install a bot within a given scope (1:1 chat, group chat, or team). If you want to get that event again you will, unfortunately, need to create a new bot channel registration and re-install the bot. 

Second, the `TeamsInfo.GetMembersAsync()` should return to you the list of all non-bot user accounts in the group chat or team. If that's not happening I can currently think of two reasons as to why. First, the bot isn't actually installed in the team that you think that it is. Second, there might be some restriction that your tenant admin has put in place that might be inhibiting the bot's ability to gather that information. 

Can you try re-installing the bot within the team and re-triggering the event and let me know the result?

**Comment by apeckstl** [comment]
@Virtual-Josh 
Thanks for the quick response! I do understand the need to reinstall each time to trigger that event, and have done so. I am getting the same result. 

I am installing the app inside the general channel of a team that I know has 4 members. To check, that same bot in the same installation in the same team, if I put the `await TeamsInfo.GetMembersAsync(turnContext, cancellationToken);` in the OnMessageActivity() method, returns the full list of 4 people. It is just on the first invocation of the OnMembersAdded() event (after deleting, reinstalling), that I only get one person returned.

**Comment by Virtual-Josh** [comment]
@apeckstl 
I think there's a tiny typo with the override with the method. The `TeamsActivityHandler` defaults back to the `ActivityHandler` behavior for some events. If the `GetMembersAsync` is working in the `OnMessage`, but not in `OnTeamsMembersAddedAsync` then I'd suspect there's some tiny error that's resulting in the base method being called vs the overridden one. 

To debug, can you add a breakpoint/console message to make sure that you're getting into the correct functions? Alternatively can you post your code so I can take a look to see if anything else jumps out at me?

**Comment by apeckstl** [comment]
Here are the relevant messages:
```
protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
{
    turnContext.Activity.RemoveRecipientMention();

    switch (turnContext.Activity.Text.Trim())
    {
        case "MentionMe":
            await MentionActivityAsync(turnContext, cancellationToken);
            break;

        case "UpdateCardAction":
            await UpdateCardActivityAsync(turnContext, cancellationToken);
            break;

        case "Delete":
            await DeleteCardActivityAsync(turnContext, cancellationToken);
            break;

        case "MessageAllMembers":
            await MessageAllMembersAsync(turnContext, cancellationToken);
            break;

        default:
            //...
            //await turnContext.SendActivityAsync(MessageFactory.Attachment(card.ToAttachment()));
            break;
    }
}

protected override async Task OnTeamsMembersAddedAsync(IList<TeamsChannelAccount> membersAdded, TeamInfo teamInfo, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
{
    IEnumerable<TeamsChannelAccount> members = await TeamsInfo.GetMembersAsync(turnContext, cancellationToken);
    foreach (var member in members)
    {
        Logger.LogInformation(member.Name);
    }
}

private async Task DeleteCardActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
{
    await turnContext.DeleteActivityAsync(turnContext.Activity.ReplyToId, cancellationToken);
}

// If you encounter permission-related errors when sending this message, see
// https://aka.ms/BotTrustServiceUrl
private async Task MessageAllMembersAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
{
    var Id = turnContext.Activity.TeamsGetChannelId();
    var serviceUrl = turnContext.Activity.ServiceUrl;
    var credentials = new MicrosoftAppCredentials(_appId, _appPassword);
    ConversationReference conversationReference = null;

    var members = await TeamsInfo.GetMembersAsync(turnContext, cancellationToken);

    foreach (var teamMember in members)
    {
        var proactiveMessage = MessageFactory.Text($"Hello {teamMember.GivenName} {teamMember.Surname}. I'm a Teams conversation bot.");

        var conversationParameters = new ConversationParameters
        {
            IsGroup = false,
            Bot = turnContext.Activity.Recipient,
            Members = new ChannelAccount[] { teamMember },
            TenantId = turnContext.Activity.Conversation.TenantId,
        };

        await ((BotFrameworkAdapter)turnContext.Adapter).CreateConversationAsync(
            Id,
            serviceUrl,
            credentials,
            conversationParameters,
            async (t1, c1) =>
            {
                conversationReference = t1.Activity.GetConversationReference();
                await ((BotFrameworkAdapter)turnContext.Adapter).ContinueConversationAsync(
                    _appId,
                    conversationReference,
                    async (t2, c2) =>
                    {
                        await t2.SendActivityAsync(proactiveMessage, c2);
                    },
                    cancellationToken);
            },
            cancellationToken);
    }

    await turnContext.SendActivityAsync(MessageFactory.Text("All messages have been sent."), cancellationToken);
}
```
The OnTeamsMembersAdded() method in my code is definitely running, as it is outputting a log message that's not there otherwise. When I send a message to the bot saying "MessageAllMembers", all of the members in the channel get a message.

**Comment by Virtual-Josh** [comment]
@apeckstl 

I ran you code (minus the logging) and the `GetMembers` is return results for me. Let's try 1 more thing. If this doesn't work I'm happy to jump on a call with you to get this sorted out.

I want to make sure nothing's wrong with the logger. If you put a break point in the `OnTeamsMembersAddedAsync` function on the line with the `foreach` loop definition and look at the `members` object under the `Results View` how many elements are in that list?

**Comment by apeckstl** [comment]
@Virtual-Josh 
I put a break point there, and it is returning just 1 person. You are saying that in your case, it returns members other than you when the bot is first added? 

I tried to add another member to the team after adding the bot, and when the breakpoint triggers again, it displays the full correct list of members. It's just when the bot is first added that I see this behavior.

**Comment by apeckstl** [comment]
Additionally, whenever I try to delete my custom app from the microsoft account, it is duplicating the entry for the app, and won't let me delete.
<img width="1200" alt="Screen Shot 2020-01-25 at 2 50 51 PM" src="https://user-images.githubusercontent.com/19844185/73127190-5b670d00-3f82-11ea-841c-5af3533394f6.png">

An error just pops up above it saying "something went wrong, please try again later".

I'm assuming this is unrelated, but do you have any advice on how to get this fixed? I can't debug properly with this happening.

**Comment by Virtual-Josh** [comment]
If you go to `Apps` you can right click and `uninstall` the bot

![image](https://user-images.githubusercontent.com/50158775/73198157-6f784f00-40e7-11ea-87da-afc3488f17e1.png)

If you have multiple bots installed it might be worth uninstalling all of them, reinstalling it, and then seeing if the issue resolves. If the issue doesn't resolve let me know and we can find a better way to unblock you.

**Comment by apeckstl** [comment]
Thanks, that resolved that issue. You mentioned that the GetMembersAsync was working for you when testing it. Does this indicate that I have something wrong with the way my team is set up? I have checked over all the permissions in our tenant and it looks fine, but I can't think of any other reason why the bot would not be picking up all the conversation members.

**Comment by Virtual-Josh** [comment]
I asked someone who would know and I don't think there's any Tenant config things that would be getting in your way. If the `GetMembersAsync` wasn't working for you I can only assume there was some weird caching issue. If `GetMembersAsync` STILL isn't working for you let me know and we can keep digging. 

For what it's worth I did have a weird interaction earlier this morning where I installed a new bot into a new team and the bot sent me a 1:1 message saying something went wrong, but it was added to the team just fine. I didn't dig into it, but the real answer might be a bug with Teams. I'm not sure.

**Comment by apeckstl** [comment]
Hi Josh,
I figured out my problem. I was misreading your original reply to this thread where you said I "need to create a new bot channel registration and re-install the bot", thinking you were saying I needed to just delete the app package from the Teams store and re-upload. I just made a new bot channel registration to start over and it worked as expected. I am so so sorry for the confusion. Thanks for all your help!
